File: lib/core/storage/tables/dose_log.dart
import 'package:drift/drift.dart';

import 'drug.dart';

/// One row per recorded dose. Used both by the schedule view and as
/// the input series for the PK calculator.
class DoseLogs extends Table {
  TextColumn get id => text()();
  TextColumn get drugId => text().references(Drugs, #id)();
  RealColumn get doseMg => real()();
  DateTimeColumn get takenAt => dateTime()();
  TextColumn get note => text().nullable()();
  DateTimeColumn get createdAt => dateTime().withDefault(currentDateAndTime)();

  @override
  Set<Column<Object>> get primaryKey => {id};
}

File: backend/app/models/dose_log.py
from datetime import datetime

from sqlalchemy import DateTime, Float, ForeignKey, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from ..core.database import Base


class DoseLog(Base):
    __tablename__ = "dose_logs"

    id: Mapped[str] = mapped_column(String(36), primary_key=True)
    drug_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("drugs.id", ondelete="CASCADE")
    )
    dose_mg: Mapped[float] = mapped_column(Float)
    taken_at: Mapped[datetime] = mapped_column(DateTime(timezone=True))
    note: Mapped[str | None] = mapped_column(Text)
